store/sqlstore/upgrade_test.go

```diff
@@ -8,34 +8,95 @@ import (
 
 	"github.com/mattermost/mattermost-server/model"
 	"github.com/mattermost/mattermost-server/store"
+	"github.com/stretchr/testify/require"
 )
 
 func TestStoreUpgrade(t *testing.T) {
 	StoreTest(t, func(t *testing.T, ss store.Store) {
-		saveSchemaVersion(ss.(*store.LayeredStore).DatabaseLayer.(SqlStore), VERSION_3_0_0)
-		UpgradeDatabase(ss.(*store.LayeredStore).DatabaseLayer.(SqlStore))
+		sqlStore := ss.(*store.LayeredStore).DatabaseLayer.(SqlStore)
 
-		saveSchemaVersion(ss.(*store.LayeredStore).DatabaseLayer.(SqlStore), "")
-		UpgradeDatabase(ss.(*store.LayeredStore).DatabaseLayer.(SqlStore))
+		t.Run("invalid currentModelVersion", func(t *testing.T) {
+			require.Panics(t, func() {
+				UpgradeDatabase(sqlStore, "notaversion")
+			})
+		})
+
+		t.Run("upgrade from invalid version", func(t *testing.T) {
+			saveSchemaVersion(sqlStore, "invalid")
+			err := UpgradeDatabase(sqlStore, "5.8.0")
+			require.EqualError(t, err, "failed to parse database schema version invalid: No Major.Minor.Patch elements found")
+			require.Equal(t, "invalid", sqlStore.GetCurrentSchemaVersion())
+		})
+
+		t.Run("upgrade from unsupported version", func(t *testing.T) {
+			saveSchemaVersion(sqlStore, "2.0.0")
+			err := UpgradeDatabase(sqlStore, "5.8.0")
+			require.EqualError(t, err, "Database schema version 2.0.0 is no longer supported. This Mattermost server supports automatic upgrades from schema version 3.0.0 through schema version 5.8.0. Please manually upgrade to at least version 3.0.0 before continuing.")
+			require.Equal(t, "2.0.0", sqlStore.GetCurrentSchemaVersion())
+		})
+
+		t.Run("upgrade from earliest supported version", func(t *testing.T) {
+			saveSchemaVersion(sqlStore, VERSION_3_0_0)
+			err := UpgradeDatabase(sqlStore, model.CurrentVersion)
+			require.NoError(t, err)
+			require.Equal(t, model.CurrentVersion, sqlStore.GetCurrentSchemaVersion())
+		})
+
+		t.Run("upgrade from no existing version", func(t *testing.T) {
+			saveSchemaVersion(sqlStore, "")
+			err := UpgradeDatabase(sqlStore, model.CurrentVersion)
+			require.NoError(t, err)
+			require.Equal(t, model.CurrentVersion, sqlStore.GetCurrentSchemaVersion())
+		})
+
+		t.Run("upgrade schema running earlier minor version", func(t *testing.T) {
+			saveSchemaVersion(sqlStore, "5.1.0")
+			err := UpgradeDatabase(sqlStore, "5.8.0")
```

Comment:
I'm specifically testing with `5.8.0` here, since the code is already at `5.10.0` and would upgrade past. Technically, I could implement logic to constrain the migrations to the desired target version, but there's currently no reason to do that outside of unit testing, so I'm limiting the changes to the real code.

In the future, we'll likely support a discrete command to upgrade/downgrade to a particular version, where this might get improved.